Solve clang-tidy warnings

See [Travis CI log](https://travis-ci.org/vgough/encfs/branches) on master branch, the clang build, to list clang-tidy warnings to correct.

The following :
```
autosprintf.cpp:38:14: warning: do not define a C-style variadic function; consider using a function parameter pack or currying instead [cert-dcl50-cpp]
autosprintf::autosprintf(const char *format, ...) {
 ^
```
Leads to :
https://www.securecoding.cert.org/confluence/display/cplusplus/DCL50-CPP.+Do+not+define+a+C-style+variadic+function

And I then found several implementations :
https://gist.github.com/sehe/3374327
https://github.com/tromey/typesafe-printf
https://github.com/fmtlib/fmt

@vgough what do you think about this ?
Would you be OK to implement a solution to fix this warning ?
(I tried but I'm stuck somewhere trying to match the old constructor with the first method above...)
Many thx 👍

May I ask some help to correctly solve the following warnings ?
@rfjakob, would you please handle them ?
Will make the build log even more clean.
Thank you very much 👍 
```
/home/travis/build/vgough/encfs/encfs/DirNode.cpp:133:7: warning: class 'RenameOp' defines a destructor and a copy constructor but does not define a copy assignment operator, a move constructor or a move assignment operator [cppcoreguidelines-special-member-functions]
class RenameOp {
      ^
/home/travis/build/vgough/encfs/encfs/NullCipher.cpp:50:7: warning: class 'NullKey' defines a destructor but does not define a copy constructor, a copy assignment operator, a move constructor or a move assignment operator [cppcoreguidelines-special-member-functions]
class NullKey : public AbstractCipherKey {
      ^
/home/travis/build/vgough/encfs/encfs/NullCipher.cpp:56:7: warning: class 'NullDestructor' defines a destructor, a copy constructor and a copy assignment operator but does not define a move constructor or a move assignment operator [cppcoreguidelines-special-member-functions]
class NullDestructor {
      ^
/home/travis/build/vgough/encfs/encfs/SSL_Cipher.cpp:277:7: warning: class 'SSLKey' defines a destructor but does not define a copy constructor, a copy assignment operator, a move constructor or a move assignment operator [cppcoreguidelines-special-member-functions]
class SSLKey : public AbstractCipherKey {
      ^
/home/travis/build/vgough/encfs/encfs/XmlReader.cpp:165:7: warning: class 'XmlNode' defines a destructor but does not define a copy constructor, a copy assignment operator, a move constructor or a move assignment operator [cppcoreguidelines-special-member-functions]
class XmlNode : virtual public XmlValue {
      ^
```

--- a/encfs/DirNode.cpp
+++ b/encfs/DirNode.cpp
@@ -145,12 +145,10 @@ class RenameOp {
   // destructor
   ~RenameOp();
 
-  // copy contructor
-  RenameOp(const RenameOp &src) = default;
-
-  RenameOp& operator=(const RenameOp& other) = delete; // copy assignment
+  RenameOp(const RenameOp &src) = delete; // copy contructor
   RenameOp(RenameOp&& other) = delete; // move constructor
-  RenameOp& operator=(RenameOp&& other) = delete ;// move assignment
+  RenameOp& operator=(const RenameOp& other) = delete; // copy assignment
+  RenameOp& operator=(RenameOp&& other) = delete; // move assignment
 
   explicit operator bool() const { return renameList != nullptr; }
 

--- a/encfs/DirNode.cpp
+++ b/encfs/DirNode.cpp
@@ -144,9 +144,15 @@ class RenameOp {
     last = renameList->begin();
   }
 
+  // destructor
+  ~RenameOp();
+
+  // copy contructor
   RenameOp(const RenameOp &src) = default;
 
-  ~RenameOp();
+  RenameOp& operator=(const RenameOp& other) = delete; // copy assignment
+  RenameOp(RenameOp&& other) = delete; // move constructor
+  RenameOp& operator=(RenameOp&& other) = delete ;// move assignment
 
   explicit operator bool() const { return renameList != nullptr; }
 

--- a/encfs/NullCipher.cpp
+++ b/encfs/NullCipher.cpp
@@ -55,9 +55,9 @@ class NullKey : public AbstractCipherKey {
   ~NullKey() override = default;
 
   NullKey(const NullKey &src) = delete; // copy constructor
-  NullKey& operator=(const NullKey& other) = delete; // copy assignment
   NullKey(NullKey&& other) = delete; // move constructor
-  NullKey& operator=(NullKey&& other) = delete ;// move assignment
+  NullKey& operator=(const NullKey& other) = delete; // copy assignment
+  NullKey& operator=(NullKey&& other) = delete; // move assignment
 };
 
 class NullDestructor {
@@ -70,13 +70,11 @@ class NullDestructor {
   // copy contructor
   NullDestructor(const NullDestructor &) = default;
 
-  // copy assignment
-  NullDestructor &operator=(const NullDestructor &) = default;
-
   // move constructor
   NullDestructor(NullDestructor &&) = default;
 
-  NullDestructor& operator=(NullDestructor&& other) = delete ;// move assignment
+  NullDestructor &operator=(const NullDestructor &) = delete; // copy assignment
+  NullDestructor& operator=(NullDestructor&& other) = delete; // move assignment
 
   void operator()(NullKey *&) {}
 };

--- a/encfs/NullCipher.cpp
+++ b/encfs/NullCipher.cpp
@@ -50,16 +50,32 @@ static bool NullCipher_registered = Cipher::Register(
 class NullKey : public AbstractCipherKey {
  public:
   NullKey() = default;
+
+  // destructor
   ~NullKey() override = default;
+
+  NullKey(const NullKey &src) = delete; // copy constructor
+  NullKey& operator=(const NullKey& other) = delete; // copy assignment
+  NullKey(NullKey&& other) = delete; // move constructor
+  NullKey& operator=(NullKey&& other) = delete ;// move assignment
 };
 
 class NullDestructor {
  public:
   NullDestructor() = default;
-  NullDestructor(const NullDestructor &) = default;
+
+  // destructor
   ~NullDestructor() = default;
 
+  // copy contructor
+  NullDestructor(const NullDestructor &) = default;
+
+  // copy assignment
   NullDestructor &operator=(const NullDestructor &) = default;
+
+  NullDestructor(NullDestructor&& other) = delete; // move constructor
+  NullDestructor& operator=(NullDestructor&& other) = delete ;// move assignment
+
   void operator()(NullKey *&) {}
 };
 std::shared_ptr<AbstractCipherKey> gNullKey(new NullKey(), NullDestructor());

--- a/encfs/NullCipher.cpp
+++ b/encfs/NullCipher.cpp
@@ -73,7 +73,9 @@ class NullDestructor {
   // copy assignment
   NullDestructor &operator=(const NullDestructor &) = default;
 
-  NullDestructor(NullDestructor&& other) = delete; // move constructor
+  // move constructor
+  NullDestructor(NullDestructor &&) = default;
+
   NullDestructor& operator=(NullDestructor&& other) = delete ;// move assignment
 
   void operator()(NullKey *&) {}

--- a/encfs/SSL_Cipher.cpp
+++ b/encfs/SSL_Cipher.cpp
@@ -293,7 +293,14 @@ class SSLKey : public AbstractCipherKey {
   HMAC_CTX *mac_ctx;
 
   SSLKey(int keySize, int ivLength);
+
+  // destructor
   ~SSLKey() override;
+
+  SSLKey(const SSLKey &src) = delete; // copy constructor
+  SSLKey& operator=(const SSLKey& other) = delete; // copy assignment
+  SSLKey(SSLKey&& other) = delete; // move constructor
+  SSLKey& operator=(SSLKey&& other) = delete ;// move assignment
 };
 
 SSLKey::SSLKey(int keySize_, int ivLength_) {

--- a/encfs/SSL_Cipher.cpp
+++ b/encfs/SSL_Cipher.cpp
@@ -298,9 +298,9 @@ class SSLKey : public AbstractCipherKey {
   ~SSLKey() override;
 
   SSLKey(const SSLKey &src) = delete; // copy constructor
-  SSLKey& operator=(const SSLKey& other) = delete; // copy assignment
   SSLKey(SSLKey&& other) = delete; // move constructor
-  SSLKey& operator=(SSLKey&& other) = delete ;// move assignment
+  SSLKey& operator=(const SSLKey& other) = delete; // copy assignment
+  SSLKey& operator=(SSLKey&& other) = delete; // move assignment
 };
 
 SSLKey::SSLKey(int keySize_, int ivLength_) {

--- a/encfs/XmlReader.cpp
+++ b/encfs/XmlReader.cpp
@@ -169,7 +169,13 @@ class XmlNode : virtual public XmlValue {
   explicit XmlNode(const tinyxml2::XMLElement *element_)
       : XmlValue(safeValueForNode(element_)), element(element_) {}
 
+  // destructor
   ~XmlNode() override = default;
+
+  XmlNode(const XmlNode &src) = delete; // copy constructor
+  XmlNode& operator=(const XmlNode& other) = delete; // copy assignment
+  XmlNode(XmlNode&& other) = delete; // move constructor
+  XmlNode& operator=(XmlNode&& other) = delete ;// move assignment
 
   XmlValuePtr find(const char *name) const override {
     if (name[0] == '@') {

--- a/encfs/XmlReader.cpp
+++ b/encfs/XmlReader.cpp
@@ -173,9 +173,9 @@ class XmlNode : virtual public XmlValue {
   ~XmlNode() override = default;
 
   XmlNode(const XmlNode &src) = delete; // copy constructor
-  XmlNode& operator=(const XmlNode& other) = delete; // copy assignment
   XmlNode(XmlNode&& other) = delete; // move constructor
-  XmlNode& operator=(XmlNode&& other) = delete ;// move assignment
+  XmlNode& operator=(const XmlNode& other) = delete; // copy assignment
+  XmlNode& operator=(XmlNode&& other) = delete; // move assignment
 
   XmlValuePtr find(const char *name) const override {
     if (name[0] == '@') {

